### PR TITLE
feat(system-api): add `call_cycles_add128_up_to`

### PR DESCRIPTION
**Pull request by sesi200**
Introduces `call_cycles_add128_up_to` to the system API to offer a way of adding as many cycles as possible to a call without trapping in the subsequent `call_perform`.
Corresponding spec MR: https://github.com/dfinity/interface-spec/pull/316
Closes [SDK-1761](https://dfinity.atlassian.net/browse/SDK-1761?atlOrigin=eyJpIjoiY2QyZTJiZDRkNGZhNGZlMWI3NzRkNTBmZmVlNzNiZTkiLCJwIjoianN3LWdpdGxhYi1pbnQifQ)

[SDK-1761]: https://dfinity.atlassian.net/browse/SDK-1761?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rs/system_api/src/sandbox_safe_system_state.rs
Comment (alin-at-dfinity) [inline rs/system_api/src/sandbox_safe_system_state.rs:547]: Nit: Please add a brief doc comment.

File: rs/system_api/src/sandbox_safe_system_state.rs
Comment (alin-at-dfinity) [inline rs/system_api/src/sandbox_safe_system_state.rs:576]: Also don't see this being used anywhere outside of `SandboxSafeSystemState`.

**Comment by dsarlis** [comment]
Adding @mraszyk for spec compliance check.

File: rs/execution_environment/src/hypervisor/tests.rs
Comment (dsarlis) [inline rs/execution_environment/src/hypervisor/tests.rs:2243]: nit: It would probably be nice to confirm that the reply additionally matches the amount requested (i.e. don't rely only on the hardcoded values provided in the test).

File: rs/execution_environment/src/hypervisor/tests.rs
Comment (dsarlis) [inline rs/execution_environment/src/hypervisor/tests.rs:2185]: nit: For these tests maybe leave a couple of comments explaining the expectations. E.g. for the first one, we should see that the entire amount is added while in the second it would only add up to what the freezing threshold dictates.

**Review by dsarlis — APPROVED** [review]

File: rs/cycles_account_manager/src/lib.rs
Comment (mraszyk) [inline rs/cycles_account_manager/src/lib.rs:405]: Why is this needed? Can't the memory also grow because the canister performs some allocation after performing the call?

File: rs/cycles_account_manager/src/lib.rs
Comment (mraszyk) [inline rs/cycles_account_manager/src/lib.rs:404]: Could this calculation be refactored into a separate function used here and in call_perform?

File: rs/system_api/src/lib.rs
Comment (mraszyk) [inline rs/system_api/src/lib.rs:2265]: Should we trace the syscall if it fails? (note that we only do `result?` on the next line)

**Review by mraszyk — APPROVED** [review]

File: rs/cycles_account_manager/src/lib.rs
Comment (sesi200) [reply rs/cycles_account_manager/src/lib.rs:404]: `call_perform` relies on intermediate results for metrics. I tried to factor it out but didn't find a solution that's worth the lines of code

File: rs/cycles_account_manager/src/lib.rs
Comment (sesi200) [reply rs/cycles_account_manager/src/lib.rs:405]: Without this line too many cycles get attached and the `call_perform` after fails. Yes, memory can grow afterwards, but I don't think we have any way of knowing by how much or if at all

File: rs/cycles_account_manager/src/lib.rs
Comment (mraszyk) [reply rs/cycles_account_manager/src/lib.rs:404]: I see, I'd leave it up to the codeowners to decide and close this conversation.

Comment (sesi200) [reply rs/cycles_account_manager/src/lib.rs:404]: Discussed with Dimitris. We'll leave it as is. Hard to factor things out, and if the calculation changes there are tests that will fail if it's only changed in one place

**Pull request merged**
